File: app/src/main/java/com/ryanrvldo/moviecatalogue/data/source/remote/RemoteRepository.java
package com.ryanrvldo.moviecatalogue.data.source.remote;

import androidx.annotation.NonNull;

import com.ryanrvldo.moviecatalogue.BuildConfig;
import com.ryanrvldo.moviecatalogue.data.source.model.Movie;
import com.ryanrvldo.moviecatalogue.data.source.model.Season;
import com.ryanrvldo.moviecatalogue.data.source.model.TvShow;
import com.ryanrvldo.moviecatalogue.data.source.remote.response.MovieResponse;
import com.ryanrvldo.moviecatalogue.data.source.remote.response.TvShowResponse;
import com.ryanrvldo.moviecatalogue.utils.EspressoIdlingResource;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RemoteRepository {

    private static RemoteRepository mInstance;
    private TmdbAPI tmdbAPI;

    private RemoteRepository(TmdbAPI tmdbAPI) {
        this.tmdbAPI = tmdbAPI;
    }

    private static OkHttpClient httpClientBuilder() {
        return new OkHttpClient.Builder()
                .readTimeout(60, TimeUnit.SECONDS)
                .connectTimeout(60, TimeUnit.SECONDS)
                .build();
    }

    public static RemoteRepository getInstance() {
        if (mInstance == null) {
            Retrofit tmdbBuilder = new Retrofit.Builder()
                    .baseUrl(BuildConfig.TMDB_BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .client(httpClientBuilder())
                    .build();
            mInstance = new RemoteRepository(tmdbBuilder.create(TmdbAPI.class));
        }
        return mInstance;
    }

    public void getMovies(String category, LoadDataCallback<MovieResponse> callback) {
        EspressoIdlingResource.increment();
        tmdbAPI.getMovies(category, BuildConfig.TMDB_API_KEY, 1)
                .enqueue(new Callback<MovieResponse>() {
                    @Override
                    public void onResponse(@NonNull Call<MovieResponse> call,
                                           @NonNull Response<MovieResponse> response) {
                        if (response.isSuccessful()) {
                            callback.onDataReceived(response.body());
                        } else {
                            callback.onDataNotAvailable();
                        }
                        EspressoIdlingResource.decrement();
                    }

                    @Override
                    public void onFailure(@NonNull Call<MovieResponse> call, @NonNull Throwable t) {
                        callback.onDataNotAvailable();
                        EspressoIdlingResource.decrement();
                    }
                });
    }

    public void searchMovies(String query, LoadDataCallback<MovieResponse> callback) {
        EspressoIdlingResource.increment();
        tmdbAPI.searchMovies(BuildConfig.TMDB_API_KEY, query)
                .enqueue(new Callback<MovieResponse>() {
                    @Override
                    public void onResponse(@NonNull Call<MovieResponse> call,
                                           @NonNull Response<MovieResponse> response) {
                        if (response.isSuccessful()) {
                            callback.onDataReceived(response.body());
                        } else {
                            callback.onDataNotAvailable();
                        }
                        EspressoIdlingResource.decrement();
                    }

                    @Override
                    public void onFailure(@NonNull Call<MovieResponse> call, @NonNull Throwable t) {
                        callback.onDataNotAvailable();
                        EspressoIdlingResource.decrement();
                    }
                });
    }

    public void getNewReleaseMovies(LoadDataCallback<MovieResponse> callback) {
        SimpleDateFormat date = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
        String currentDate = date.format(new Date());
        tmdbAPI.getNewReleaseMovies(BuildConfig.TMDB_API_KEY, currentDate, currentDate)
                .enqueue(new Callback<MovieResponse>() {
                    @Override
                    public void onResponse(@NonNull Call<MovieResponse> call,
                                           @NonNull Response<MovieResponse> response) {
                        if (response.isSuccessful()) {
                            if (response.body() != null && response.body().getMovieItems() != null) {
                                callback.onDataReceived(response.body());
                            } else {
                                callback.onDataNotAvailable();
                            }
                        } else {
                            callback.onDataNotAvailable();
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<MovieResponse> call, @NonNull Throwable t) {
                        callback.onDataNotAvailable();
                    }
                });
    }

    public void getMovieDetail(int id, LoadDataCallback<Movie> callback) {
        EspressoIdlingResource.increment();
        tmdbAPI.
                getMovie(id, BuildConfig.TMDB_API_KEY, "images,videos,credits,similar")
                .enqueue(new Callback<Movie>() {
                    @Override
                    public void onResponse(@NonNull Call<Movie> call,
                                           @NonNull Response<Movie> response) {
                        if (response.isSuccessful()) {
                            callback.onDataReceived(response.body());
                        } else {
                            callback.onDataNotAvailable();
                        }
                        EspressoIdlingResource.decrement();
                    }

                    @Override
                    public void onFailure(@NonNull Call<Movie> call, @NonNull Throwable t) {
                        callback.onDataNotAvailable();
                        EspressoIdlingResource.decrement();
                    }
                });
    }

    public void getTvShows(String category, LoadDataCallback<TvShowResponse> callback) {
        EspressoIdlingResource.increment();
        tmdbAPI.getTvShows(category, BuildConfig.TMDB_API_KEY, 1)
                .enqueue(new Callback<TvShowResponse>() {
                    @Override
                    public void onResponse(@NonNull Call<TvShowResponse> call,
                                           @NonNull Response<TvShowResponse> response) {
                        if (response.isSuccessful()) {
                            callback.onDataReceived(response.body());
                        } else {
                            callback.onDataNotAvailable();
                        }
                        EspressoIdlingResource.decrement();
                    }

                    @Override
                    public void onFailure(@NonNull Call<TvShowResponse> call, @NonNull Throwable t) {
                        callback.onDataNotAvailable();
                        EspressoIdlingResource.decrement();
                    }
                });
    }

    public void searchTvShows(String query, LoadDataCallback<TvShowResponse> callback) {
        EspressoIdlingResource.increment();
        tmdbAPI.searchTvShows(BuildConfig.TMDB_API_KEY, query)
                .enqueue(new Callback<TvShowResponse>() {
                    @Override
                    public void onResponse(@NonNull Call<TvShowResponse> call,
                                           @NonNull Response<TvShowResponse> response) {
                        if (response.isSuccessful()) {
                            callback.onDataReceived(response.body());
                        } else {
                            callback.onDataNotAvailable();
                        }
                        EspressoIdlingResource.decrement();
                    }

                    @Override
                    public void onFailure(@NonNull Call<TvShowResponse> call, @NonNull Throwable t) {
                        callback.onDataNotAvailable();
                        EspressoIdlingResource.decrement();
                    }
                });
    }


    public void getNewReleaseTvShows(LoadDataCallback<TvShowResponse> callback) {
        SimpleDateFormat date = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
        String currentDate = date.format(new Date());
        tmdbAPI.getNewReleaseTVShow(BuildConfig.TMDB_API_KEY, currentDate, currentDate)
                .enqueue(new Callback<TvShowResponse>() {
                    @Override
                    public void onResponse(@NonNull Call<TvShowResponse> call,
                                           @NonNull Response<TvShowResponse> response) {
                        if (response.isSuccessful()) {
                            if (response.body() != null && response.body().getTvShowItems() != null) {
                                callback.onDataReceived(response.body());
                            } else {
                                callback.onDataNotAvailable();
                            }
                        } else {
                            callback.onDataNotAvailable();
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<TvShowResponse> call, @NonNull Throwable t) {
                        callback.onDataNotAvailable();
                    }
                });
    }

    public void getTvDetail(int id, LoadDataCallback<TvShow> callback) {
        EspressoIdlingResource.increment();
        tmdbAPI.getTvShow(id, BuildConfig.TMDB_API_KEY, "images,content_ratings,videos,credits,similar")
                .enqueue(new Callback<TvShow>() {
                    @Override
                    public void onResponse(@NonNull Call<TvShow> call,
                                           @NonNull Response<TvShow> response) {
                        if (response.isSuccessful()) {
                            callback.onDataReceived(response.body());
                        } else {
                            callback.onDataNotAvailable();
                        }
                        EspressoIdlingResource.decrement();
                    }

                    @Override
                    public void onFailure(@NonNull Call<TvShow> call, @NonNull Throwable t) {
                        callback.onDataNotAvailable();
                        EspressoIdlingResource.decrement();
                    }
                });
    }

    public void getSeasonDetail(int tvId, int seasonNumber, LoadDataCallback<Season> callback) {
        EspressoIdlingResource.increment();
        tmdbAPI.getSeasonTvDetail(tvId, seasonNumber, BuildConfig.TMDB_API_KEY)
                .enqueue(new Callback<Season>() {
                    @Override
                    public void onResponse(@NonNull Call<Season> call,
                                           @NonNull Response<Season> response) {
                        if (response.isSuccessful()) {
                            if (response.body() != null && response.body().getEpisodes() != null) {
                                callback.onDataReceived(response.body());
                            } else {
                                callback.onDataNotAvailable();
                            }
                        } else {
                            callback.onDataNotAvailable();
                        }
                        EspressoIdlingResource.decrement();
                    }

                    @Override
                    public void onFailure(@NonNull Call<Season> call, @NonNull Throwable t) {
                        callback.onDataNotAvailable();
                        EspressoIdlingResource.decrement();
                    }
                });
    }
}

File: app/src/main/kotlin/com/ryanrvldo/moviecatalogue/data/remote/RemoteRepository.kt
package com.ryanrvldo.moviecatalogue.data.remote

import com.ryanrvldo.moviecatalogue.data.model.Movie
import com.ryanrvldo.moviecatalogue.data.model.Season
import com.ryanrvldo.moviecatalogue.data.model.TvShow
import com.ryanrvldo.moviecatalogue.data.remote.response.MovieResponse
import com.ryanrvldo.moviecatalogue.data.remote.response.TvShowResponse
import com.ryanrvldo.moviecatalogue.data.vo.Resource
import com.ryanrvldo.moviecatalogue.utils.EspressoIdlingResource
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject

class RemoteRepository @Inject constructor(
    private val tmdbAPI: TmdbAPI
) : ResponseHelper() {

    suspend fun getMovies(category: String): Resource<MovieResponse> = getResponseResult {
        tmdbAPI.getMovieResponse(category)
    }

    suspend fun getMovieDetails(id: Int): Resource<Movie> = getResponseResult {
        tmdbAPI.getMovieDetails(id)
    }

    suspend fun searchMovies(query: String) = getResponseResult {
        tmdbAPI.searchMovies(query)
    }

    fun getNewReleaseMovies(callback: LoadDataCallback<MovieResponse>) {
        val date = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
        val currentDate = date.format(Date())
        tmdbAPI.getNewReleaseMovies(currentDate, currentDate)
            .enqueue(object : Callback<MovieResponse?> {
                override fun onResponse(
                    call: Call<MovieResponse?>,
                    response: Response<MovieResponse?>
                ) {
                    if (response.isSuccessful) {
                        response.body()?.let { callback.onDataReceived(it) }
                    } else {
                        callback.onDataNotAvailable()
                    }
                }

                override fun onFailure(
                    call: Call<MovieResponse?>,
                    t: Throwable
                ) {
                    callback.onDataNotAvailable()
                }
            })
    }

    suspend fun getTvShows(category: String): Resource<TvShowResponse> = getResponseResult {
        tmdbAPI.getTvShowResponse(category)
    }

    suspend fun getTvDetail(id: Int): Resource<TvShow> = getResponseResult {
        tmdbAPI.getTvShowDetails(id)
    }

    suspend fun searchTvShows(query: String): Resource<TvShowResponse> = getResponseResult {
        tmdbAPI.searchTvShows(query)
    }

    fun getNewReleaseTvShows(callback: LoadDataCallback<TvShowResponse>) {
        val date = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
        val currentDate = date.format(Date())
        tmdbAPI.getNewReleaseTVShow(currentDate, currentDate)
            .enqueue(object : Callback<TvShowResponse?> {
                override fun onResponse(
                    call: Call<TvShowResponse?>,
                    response: Response<TvShowResponse?>
                ) {
                    if (response.isSuccessful) {
                        response.body()?.let { callback.onDataReceived(it) }
                    } else {
                        callback.onDataNotAvailable()
                    }
                }

                override fun onFailure(
                    call: Call<TvShowResponse?>,
                    t: Throwable
                ) {
                    callback.onDataNotAvailable()
                }
            })
    }

    fun getSeasonDetail(
        tvId: Int, seasonNumber: Int, callback: LoadDataCallback<Season>
    ) {
        EspressoIdlingResource.increment()
        tmdbAPI.getSeasonTvDetail(tvId, seasonNumber)
            .enqueue(object : Callback<Season?> {
                override fun onResponse(
                    call: Call<Season?>,
                    response: Response<Season?>
                ) {
                    if (response.isSuccessful) {
                        response.body()?.let { callback.onDataReceived(it) }
                    } else {
                        callback.onDataNotAvailable()
                    }
                    EspressoIdlingResource.decrement()
                }

                override fun onFailure(
                    call: Call<Season?>,
                    t: Throwable
                ) {
                    callback.onDataNotAvailable()
                    EspressoIdlingResource.decrement()
                }
            })
    }

}

